WcaOnRails/app/models/concerns/taggable.rb

```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+# This module can be included by a model willing to have tags.
+# The including model must have a has_many relationship to their tags table named "item_tags"
```

Comment:
I was actually looking to do something like this!
Shouldn't we be only doing the public_send?
Would the `@tags_association` be reloaded if we do something like `incident.reload`?

Reply:
I believe it would still work
```ruby
jonatanklosko@mockingjay:~/projects/worldcubeassociation.org/WcaOnRails (master) $ rails c --sandbox
irb(main):001:0> person = Person.first
  Person Load (0.9ms)  SELECT  `rails_persons`.* FROM `rails_persons` ORDER BY `rails_persons`.`id` ASC LIMIT 1
=> #<Person id: 1, wca_id: "1982BORS01", subId: 1, name: "Jozsef Borsos", countryId: "Serbia", gender: "m", year: 1954, month: 12, day: 4, comments: "">
irb(main):002:0> competitions_association = person.competitions
  Competition Load (1.1ms)  SELECT  DISTINCT `Competitions`.* FROM `Competitions` INNER JOIN `Results` ON `Competitions`.`id` = `Results`.`competitionId` WHERE `Results`.`personId` = '1982BORS01' LIMIT 11
=> #<ActiveRecord::Associations::CollectionProxy [#<Competition id: "WC1982", name: "World Rubik's Cube Championship 1982", cityName: "Budapest", countryId: "Hungary", information: "", year: 1982, month: 6, day: 5, endMonth: 6, endDay: 5, venue: "", venueAddress: "", venueDetails: "", external_website: "", cellName: "World Championship 1982", showAtAll: true, latitude: 47498403, longitude: 19040759, isConfirmed: true, contact: "", remarks: "remarks to the board here", registration_open: nil, registration_close: nil, use_wca_registration: false, guests_enabled: true, results_posted_at: "1982-06-05 00:00:00", results_nag_sent_at: "2016-09-04 14:01:15", generate_website: nil, announced_at: nil, base_entry_fee_lowest_denomination: 0, currency_code: "USD", endYear: 1982, connected_stripe_account_id: nil, start_date: "1982-06-05", end_date: "1982-06-05", enable_donations: nil, competitor_limit_enabled: false, competitor_limit: nil, competitor_limit_reason: nil>]>
irb(main):003:0> person.competitions.destroy_all
  Competition Load (1.7ms)  SELECT DISTINCT `Competitions`.* FROM `Competitions` INNER JOIN `Results` ON `Competitions`.`id` = `Results`.`competitionId` WHERE `Results`.`personId` = '1982BORS01'
   (0.3ms)  SAVEPOINT active_record_1
  Result Load (0.6ms)  SELECT `Results`.* FROM `Results` WHERE `Results`.`personId` = '1982BORS01' AND `Results`.`competitionId` = 'WC1982'
  SQL (0.8ms)  DELETE FROM `Results` WHERE `Results`.`id` = 83474
   (0.3ms)  RELEASE SAVEPOINT active_record_1
=> [#<Competition id: "WC1982", name: "World Rubik's Cube Championship 1982", cityName: "Budapest", countryId: "Hungary", information: "", year: 1982, month: 6, day: 5, endMonth: 6, endDay: 5, venue: "", venueAddress: "", venueDetails: "", external_website: "", cellName: "World Championship 1982", showAtAll: true, latitude: 47498403, longitude: 19040759, isConfirmed: true, contact: "", remarks: "remarks to the board here", registration_open: nil, registration_close: nil, use_wca_registration: false, guests_enabled: true, results_posted_at: "1982-06-05 00:00:00", results_nag_sent_at: "2016-09-04 14:01:15", generate_website: nil, announced_at: nil, base_entry_fee_lowest_denomination: 0, currency_code: "USD", endYear: 1982, connected_stripe_account_id: nil, start_date: "1982-06-05", end_date: "1982-06-05", enable_donations: nil, competitor_limit_enabled: false, competitor_limit: nil, competitor_limit_reason: nil>]
irb(main):004:0> competitions_association
=> #<ActiveRecord::Associations::CollectionProxy []>
irb(main):005:0> competitions_association === person.competitions
=> true
irb(main):006:0> person.reload
  Person Load (1.0ms)  SELECT  `rails_persons`.* FROM `rails_persons` WHERE `rails_persons`.`id` = 1 LIMIT 1
=> #<Person id: 1, wca_id: "1982BORS01", subId: 1, name: "Jozsef Borsos", countryId: "Serbia", gender: "m", year: 1954, month: 12, day: 4, comments: "">
irb(main):007:0> competitions_association === person.competitions
  Competition Load (2.1ms)  SELECT DISTINCT `Competitions`.* FROM `Competitions` INNER JOIN `Results` ON `Competitions`.`id` = `Results`.`competitionId` WHERE `Results`.`personId` = '1982BORS01'
=> true
```

Reply:
Actually the `||=` is responsible for [this](https://travis-ci.org/thewca/worldcubeassociation.org/builds/276062700#L3357) test failure!
I don't get why yet, but during the test we do [this](https://github.com/thewca/worldcubeassociation.org/blob/948d5e70114911f538ae6ad67cf724c8ac7ad6e5/WcaOnRails/spec/models/post_spec.rb#L41), which goes to `Taggable`'s `before_validation` and this happens:
```
[24, 33] in /home/fifi/Documents/cube/wca/worldcubeassociation.org/WcaOnRails/app/models/concerns/taggable.rb
   24:       tags.split(",")
   25:     end
   26: 
   27:     before_validation do
   28:       byebug
=> 29:       tags_array.each do |tag|
   30:         item_tags.find_or_initialize_by(tag: tag)
   31:       end
   32:       byebug
   33: 
(byebug) tags_array
["wdc"]
(byebug) item_tags
#<ActiveRecord::Associations::CollectionProxy [#<PostTag id: 1, post_id: 1, tag: "wdc">, #<PostTag id: 2, post_id: 1, tag: "test-with-hyphens">]>
(byebug) c

[29, 38] in /home/fifi/Documents/cube/wca/worldcubeassociation.org/WcaOnRails/app/models/concerns/taggable.rb
   29:       tags_array.each do |tag|
   30:         item_tags.find_or_initialize_by(tag: tag)
   31:       end
   32:       byebug
   33: 
=> 34:       item_tags.each do |item_tag|
   35:         item_tag.mark_for_destruction unless tags_array.include?(item_tag.tag)
   36:       end
   37:       byebug
   38:     end
(byebug) item_tags
#<ActiveRecord::Associations::CollectionProxy [#<PostTag id: 1, post_id: 1, tag: "wdc">, #<PostTag id: 2, post_id: 1, tag: "test-with-hyphens">, #<PostTag id: nil, post_id: 1, tag: "wdc">]>
```

Clearly the `find_or_initialize_by` doesn't work as expected in this case, whereas removing the `||=` makes the test pass...